src/mesh/mesh.h

```diff
@@ -357,9 +342,58 @@ class Mesh {
     /// Maximal distance of observe point from Mesh relative to its size
     double global_observe_radius() const;
 
-    /// Number of elements read from input.
-    unsigned int n_all_input_elements_;
+    /// Initialize element_vec_, set size and reset counters of boundary and bulk elements.
+    void init_element_vector(unsigned int size);
 
+    /// Returns range (of indexes in element_vector_) of boundary or bulk elements
+    inline pair<unsigned int, unsigned int> get_element_range(bool boundary=false) const {
+    	if (boundary) return make_pair<unsigned int, unsigned int>(element_vec_.size()-boundary_size_, element_vec_.size());
+    	else return make_pair<unsigned int, unsigned int>(0, bulk_size_+0);
+    }
+
+    /// Returns count of boundary or bulk elements
+    inline unsigned int n_elements(bool boundary=false) const {
+    	if (boundary) return boundary_size_;
+    	else return bulk_size_;
+    }
+
+    /// Begin iterator to bulk part of element vector
+    inline ElementIterator bulk_begin() {
+    	return element_vec_.begin();
+    }
+
+    /// End iterator to bulk part of element vector
+    inline ElementIterator bulk_end() {
+    	return element_vec_.begin() + bulk_size_;
+    }
+
+    /// Begin iterator to boundary part of element vector
+    inline ElementIterator boundary_begin() {
+    	return element_vec_.end() - boundary_size_;
+    }
+
+    /// End iterator to boundary part of element vector
+    inline ElementIterator boundary_end() {
+    	return element_vec_.end();
+    }
+
+    // For each node the vector contains a list of elements that use this node
+    vector<vector<unsigned int> > node_elements_;
+
+    /// For element of given elem_id returns index in element_vec_ or (-1) if element doesn't exist.
+    inline int elem_index(int elem_id) const
```

Comment:
Make a class for bi-dir map, make its instance, e.g. Mesh::element_ids. Kid of this method would be method of that bi-dir map class.